Add script to check for websites validity

tupaschoal opened the pull request:
This is a first attempt to ensure that we always have valid links on our JSON. I have divided it into several commits as I was cleaning up the JSON. 
- Most of them are related to websites that really ceased to exist
- Some are fixing broken links on existing websites
- Some are fixes to the URL format in JSON or adding URLS to entries without ones
- Then there is the commit itself.

I didn't make it error out just like the other, because I still have to figure out two things:
- Links to restricted sections of a website that is still valid, which require login and thus returns 404
- Some websites (actually just `premera.com`) spuriously timing out.

Nightfirecat commented:
Looks like a good start. I agree with your assessments of the issues to solve with it. That's the same reason that we run htmlproofer in cibuild without the `LinkCheck` routine--it ran into the same problems.

```
bundle exec htmlproofer ./_site --checks-to-ignore 'LinkCheck'
```

tupaschoal commented:
I changed the script a bit. Seems like using `get_response` instead of `request_head` made half of the 404 errors go away.
The other half I squashed when I stopped using `URI.parse` as I found out it was actually making valid links invalid.

Now we're just bound to the spuriousness of websites sometimes taking long to answer or becoming offline, but I think that's pretty ok. I wouldn't make it error out anyway, so that we don't blame PRs for things that are not their fault, as now we have a manageable list on the log to look into (currently 0 + any timeouts, which are 1-2)

tupaschoal commented:
I intend to merge this as is. Any other thoughts @Nightfirecat ?

Nightfirecat commented:
I'll give this another look over tonight. As-is, this PR would not affect our CI at all. Is that intentional? I think even if we're using this script only to output info-level messages for unreachable sites (and not fail the build), we should be running it consistently rather than leaving it as a static asset which isn't run on a regular basis.

tupaschoal commented:
The commented out part is temporary, I intend to enable it for the CI as soon as we merge

Nightfirecat commented:
I'm a little confused. In that case, why not uncomment it for this branch? Travis runs CI on PRs, so we would get the same result here as we would on master.

tupaschoal commented:
I got an error when uncommenting once, but it was an earlier version of the script. Didn't want to do many "fix ci" commits, as it would pollute the history and I also didn't want this PR to get squashed, because each commit is doing a unique thing on its own. Don't know if I was actually clearer now hahahaha

Nightfirecat commented:
I'd like to rebase the history pretty significantly before we merge this anyway, since there's already a number of "Fix ci" or "Fix bug" type commits as-is

tupaschoal commented:
Ok, feel free to rebase then. I made it work, finally 🎉

Nightfirecat commented:
@tupaschoal I did some quick-and-dirty changes. Mainly to keep responsibilities separate between the two scripts. validate_json.rb should be responsible for validating all aspects of JSON files, and ping_websites.rb should expect valid JSON in place before being run. lmk if this looks fine then I'll rebase and re-push for final review tomorrow

tupaschoal commented on script/ping_websites.rb:
```diff
@@ -44,46 +44,32 @@ def url_exist(name, url_string)
     elsif res.code == "404"
-        # Some webpages return 404 even though they properly redirect to login pages
-        STDERR.puts "Entry: #{name} returned #{res.to_s}"
+        # Some pages 404 even though they properly redirect to login pages
```
I think we can remove this comment, because after my latest change it doesn't happen anymore

Nightfirecat replied:
I assume that's only because sites.json doesn't have any pages that give HTTP 404 responses? Or am I missing something here?

tupaschoal replied:
Before c609a24d there were around a dozen websites falling, but they were actually alive. Afterwards, these false positives disappeared

Nightfirecat replied:
Does that make this comment invalid? Seems like it's worth keeping in case other sites end up going through this code path in the future.

tupaschoal commented:
>  validate_json.rb should be responsible for validating all aspects of JSON files, and ping_websites.rb should expect valid JSON in place before being run

I laid on my bed thinking I needed to do this change hehe.

I agree with the changes, just a minor comment and I think we're ready to rebase and go

tupaschoal commented:
@Nightfirecat I rebased all the commits, squashing some, and it seems mostly fine for me now, can you please take a look so we can merge?

Thanks!

tupaschoal commented:
I don't like the idea of merging all three as "Remove sites that cause script errors", they did indeed, but because the website was offline and now we deal with this properly. As for merging, I'm ok with joining 8189c86a3ad1870e999ebe84963c28c77d83c4dd and ac1b17aa9ea5d650257ed3a6684c85f7f1d46666, which are removals, and keeping e18dbc8e77456f438ec74ba367d63d67b8d87898 on its own, which is just altering an entry.

If you are able to deal with those changes, feel free to do so and push later.

Nightfirecat commented:
I mostly meant that I don't know what "Remove entries yielding Open/ReadTimeout" means realistically since URLs like `http://www.mylife.com/showDeleteAccount.do` redirect to a valid login page, but are removed in the commit without further explanation.

tupaschoal commented:
> I mostly meant that I don't know what "Remove entries yielding Open/ReadTimeout" means realistically since URLs like http://www.mylife.com/showDeleteAccount.do redirect to a valid login page, but are removed in the commit without further explanation.

It means that when trying to open the URL they timed-out without receiving an answer. I double checked all the removals on these commits. `mylife.com` opened (the only one), it might have been offline on the day so I reverted it back.

I also squashed the two commits about timeouts and removed the comment on the script that was outdated. It looks ready to go now.

Nightfirecat commented on _data/sites.json:
```diff
@@ -6154,16 +5984,6 @@
 
-    {
-        "name": "Riffle",
-        "url": "https://riffle.uservoice.com/knowledgebase/articles/244346-delete-your-account",
```
This URL is gone, but Riffle still exists and has an account deactivation/deletion process here: https://help.realnames.com/hc/en-us/articles/202814089-Deactivate-your-Email-Address

tupaschoal commented:
Looks ready to go (again) @Nightfirecat !

tupaschoal commented:
Pocoyo club comes and goes since I started developing this, I'll give it some days  and remove if still applicable

Nightfirecat commented:
Hm, ok then. 👍

tupaschoal commented:
See https://travis-ci.org/jdm-contrib/jdm/builds/590011232
